### PR TITLE
bpo-1635741: convert unicode ucd type to heap type

### PR DESCRIPTION
**Pull request by koubaa**
change taken from https://github.com/python/cpython/pull/22145

https://bugs.python.org/issue1635741

### CODE REVIEW

**Comment by koubaa** [comment]
@vstinner  please review

File: Modules/unicodedata.c [hunk 1]
```diff
@@ -1387,50 +1387,26 @@ static PyMethodDef unicodedata_functions[] = {
...
-        0,                      /*tp_descr_get*/
-        0,                      /*tp_descr_set*/
-        0,                      /*tp_dictoffset*/
-        0,                      /*tp_init*/
-        0,                      /*tp_alloc*/
-        0,                      /*tp_new*/
-        0,                      /*tp_free*/
-        0,                      /*tp_is_gc*/
+static void
+ucd_dealloc(PreviousDBVersion *self) {
```
Comment (vstinner) [inline Modules/unicodedata.c]: PEP 7 (coding style):

```suggestion
static void
ucd_dealloc(PreviousDBVersion *self)
{
```

File: Modules/unicodedata.c [hunk 2]
```diff
@@ -1387,50 +1387,26 @@ static PyMethodDef unicodedata_functions[] = {
...
+    PyObject_Del(self);
+    Py_DECREF(tp);
+}
+
+static PyType_Slot unicodedata_ucd_type_slots[] = {
+    {Py_tp_dealloc, ucd_dealloc},
+    {Py_tp_getattro, PyObject_GenericGetAttr},
+    {Py_tp_methods, unicodedata_functions},
+    {Py_tp_members, DB_members},
+    {0,0}
```
Comment (vstinner) [inline Modules/unicodedata.c]: nitpick:

```suggestion
    {0, 0}
```

File: Modules/unicodedata.c [hunk 3]
```diff
@@ -1457,28 +1433,45 @@ static struct PyModuleDef unicodedatamodule = {
...
+    }
 
-    PyModule_AddStringConstant(m, "unidata_version", UNIDATA_VERSION);
-    Py_INCREF(&UCD_Type);
-    PyModule_AddObject(m, "UCD", (PyObject*)&UCD_Type);
+    if (UCD_Type == NULL) {
+        UCD_Type = (PyTypeObject *)PyType_FromSpec(
+            &unicodedata_ucd_type_spec);
+        if (UCD_Type == NULL) {
+            return NULL;
```
Comment (vstinner) [inline Modules/unicodedata.c]: It seems like you want to refactor this function. In this case, please fix a memory leak: Py_DECREF(mod) is needed on error. Please add an "error:" label, and use "goto error".

File: Modules/unicodedata.c [hunk 4]
```diff
@@ -1457,28 +1433,45 @@ static struct PyModuleDef unicodedatamodule = {
...
+        }
+    }
+
+    PyModule_AddStringConstant(mod, "unidata_version", UNIDATA_VERSION);
```
Comment (vstinner) [inline Modules/unicodedata.c]: Please add error handling. Same remark for other "add" calls.

File: Modules/unicodedata.c [hunk 5]
```diff
@@ -1457,28 +1433,45 @@ static struct PyModuleDef unicodedatamodule = {
...
+    Py_INCREF(UCD_Type);
+    PyModule_AddObject(mod, "UCD", (PyObject*)UCD_Type);
```
Comment (vstinner) [inline Modules/unicodedata.c]: PyModule_AddType() can be used here, it avoids the INCREF/DECREF dance.

**Comment by vstinner** [comment]
While this change is correct, I'm not excited by leaking a new heap type at Python exit.

Would you mind to attempt to add a module state to the module, and pass to all functions which use the UCD_Type? (without converting the UCD_Type to a heap type) The module state can be an empty structure, or add a "int dummy" just to make it non-empty.

**Comment by koubaa** [comment]
> While this change is correct, I'm not excited by leaking a new heap type at Python exit.
> 
> Would you mind to attempt to add a module state to the module, and pass to all functions which use the UCD_Type? (without converting the UCD_Type to a heap type) The module state can be an empty structure, or add a "int dummy" just to make it non-empty.

@vstinner This is going to a large set of changes, almost as large as the original PR, because I have to introduce an additional layer for all methods which are used by both the module and the type.  I don't see how this will help with the leak

**Comment by vstinner** [comment]
This PR is outdated, unicodedata got many changes in the meanwhile. I proposed one approach to convert unicodedata to multi-phase init in https://bugs.python.org/issue42157 I close this PR. Once PR #22990 will be merged, I will propose a PR to finally convert the module to multi-phase init. Sorry for the misunderstanding, but this extension module is way more complex than other extensions, and I didn't spot all corner cases at the first review. See the bpo for the list of all issues and my proposal.

**Comment by koubaa** [comment]
@vstinner no problem, I agree it is complex.

**Pull request closed without merging**